### DISCUSSION TITLE
Intelligent API changes Beta 01

### DISCUSSION THREAD

**Discussion by BottlecapDave** [BottlecapDave/HomeAssistant-OctopusEnergy#921] [General]
This release includes everything in v11.2.0 as well as the updates outlined in https://github.com/BottlecapDave/HomeAssistant-OctopusEnergy/issues/853 which includes

* fix: Migrated planned and completed dispatches request to use new fields (5 minutes dev time)
* fix: Migrated retrieving intelligent device to new endpoint (1 hour dev time)
* fix: Migrated retrieving intelligent settings to new endpoints (1 hour dev time)

If you are on intelligent, it would be good if you could try the beta and report any issues in the linked discussion thread.

<hr /><em>This discussion was created from the release <a href='https://github.com/BottlecapDave/HomeAssistant-OctopusEnergy/releases/tag/intelligent-api-changes-beta-01'>Intelligent API changes Beta 01</a>.</em>

**Comment by Cameleon99** [comment]
@BottlecapDave having a few issues (as per TwiX): 

Trying to set Intelligent Charge Limit (% to add) or Ready Time results in the following error:

![Screenshot 2024-06-16 at 19 45 20](https://github.com/BottlecapDave/HomeAssistant-OctopusEnergy/assets/27414924/674ba9b4-723a-4568-baaf-fe0ac3817bc9)


Trying to set Bump charge or Enable/disable smart charge has no effect or error

**Reply by BottlecapDave** [reply]
Thanks for raising. A fix should be available in https://github.com/BottlecapDave/HomeAssistant-OctopusEnergy/releases/tag/intelligent-api-changes-beta-02

**Reply by Cameleon99** [reply]
Looks to have sorted it, I can change the 'charge to add' and 'ready by time' now with no errors reported and updated values appear on the Octopus app. Top stuff.

**Reply by BottlecapDave** [reply]
Excellent. May I ask if vehicle battery size or charger power are supplied correctly as attributes in the dispatching sensor?

**Reply by Cameleon99** [reply]
Looks right to me (on my phone atm). Charger is a Zappi (single phase 7.2kW) and car is a Zoe ZE50 R135 (52-54kWh battery depending who you ask)

![image](https://github.com/BottlecapDave/HomeAssistant-OctopusEnergy/assets/27414924/e01512cc-8004-4f92-ae2c-a8443ffc44f2)

**Reply by BottlecapDave** [reply]
hmm I didn't expect both to be set in the beta (`intelligent-api-changes-beta-02`) due to the data being returned only being either a vehicle or a battery (but not both).

**Reply by Cameleon99** [reply]
When you enrol, you pick either a car or a charger as the device to control. When you pick a compatible charger (like the Zappi), it also asks you what car you have next, so it can establish the size of the battery. So when you request to add say 10%, it knows how long to run the charger for to get the right kWh into the battery. So it makes sense to me it knows them both, but no idea what's meant to be on the API. 

I need to get the GraphQL API into my Postman so I can have a play with it. Since you had everything I need in the integration I've not needed to play with it so far apart from when they were trialling the Skoda Enyaq before they supported the Zappi and before I had the Zoe.

**Reply by Cameleon99** [reply]
This is what I get from registeredKrakenFlexDevice, gives car and charge point, so battery size and speed of EVSE.

```
{
    "data": {
        "registeredKrakenflexDevice": {
            "krakenflexDeviceId": "00000000-0000-0000-0000-00000000a0a0",
            "provider": "MYENERGI",
            "vehicleMake": "Renault",
            "vehicleModel": "Zoe ZE50 R135",
            "vehicleBatterySizeInKwh": "52.00",
            "chargePointMake": "Myenergi",
            "chargePointModel": "Zappi smart EV",
            "chargePointPowerInKw": "7.400",
            "status": "Live",
            "suspended": false,
            "hasToken": true,
            "createdAt": "2023-09-20T09:58:55.955828+00:00",
            "testDispatchFailureReason": "NONE"
        }
    }
}
```

Let me know if you want me to hit up any specific endpoints on the GraphQL API as I have it in Postman now.

**Reply by BottlecapDave** [reply]
The registeredKrakenflexDevice endpoint is what is currently being used, however they've deprecated the use of that endpoint, which is what this beta is implementing. You can see the query at https://github.com/BottlecapDave/HomeAssistant-OctopusEnergy/blob/91282c007b2f1548deaf05e5147f236e158b8c28/custom_components/octopus_energy/api_client/__init__.py#L133, where I have to retrieve devices, filter them down to electric vehicles and then match them up with their corresponding model. I'm wondering if there is multiple devices present, one for the vehicle and one for the charger.

**Reply by Cameleon99** [reply]
That's returning a big old payload, I assume you just want this bit? 

```
"devices": [
            {
                "id": "00000000-0000-0000-0000-00000000a0a0",
                "provider": "MYENERGI",
                "deviceType": "ELECTRIC_VEHICLES",
                "status": {
                    "current": "LIVE"
                },
                "__typename": "SmartFlexChargePoint",
                "make": "Myenergi",
                "model": "Zappi smart EV"
            }
        ]
```

[full_resp.json](https://github.com/user-attachments/files/15898647/full_resp.json)

**Reply by Cameleon99** [reply]
OK I think it must not have updated last time. On Beta 2 now, just has EVSE details now. 

<img width="417" alt="image" src="https://github.com/BottlecapDave/HomeAssistant-OctopusEnergy/assets/27414924/6870d77c-919d-46dc-8c04-79483cc0608a">

OE does know I have a Zoe, so should be able to get the details from the API somewhere I assume.

**Reply by BottlecapDave** [reply]
If it was to be exposed, I would have potentially expected multiple "devices"; one for the electric vehicle and one for the battery. I'll ask in the forums. But thanks for letting me know and helping out.

Side question, can you now successfully update your settings with the current beta version?

**Reply by Cameleon99** [reply]
@BottlecapDave not sure when this started happening, seems its all broken :(

One beta 2

```
2024-06-21 12:07:21.490 ERROR (MainThread) [custom_components.octopus_energy.coordinators.intelligent_settings] Unexpected error fetching intelligent_settings_A-99AAAA99 data
Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/helpers/update_coordinator.py", line 312, in _async_refresh
    self.data = await self._async_update_data()
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/helpers/update_coordinator.py", line 268, in _async_update_data
    return await self.update_method()
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/config/custom_components/octopus_energy/coordinators/intelligent_settings.py", line 98, in async_update_intelligent_settings_data
    hass.data[DOMAIN][account_id][DATA_INTELLIGENT_SETTINGS] = await async_refresh_intelligent_settings(
                                                               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/config/custom_components/octopus_energy/coordinators/intelligent_settings.py", line 51, in async_refresh_intelligent_settings
    settings = await client.async_get_intelligent_settings(account_id, device_id)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/config/custom_components/octopus_energy/api_client/__init__.py", line 1059, in async_get_intelligent_settings
    self.__ready_time_to_time__(charging_preferences["weekdayTargetTime"])
  File "/config/custom_components/octopus_energy/api_client/__init__.py", line 1079, in __ready_time_to_time__
    raise Exception(f"Unexpected number of parts in '{time_str}'")
Exception: Unexpected number of parts in '05:30:00'
```

**Reply by Cameleon99** [reply]
Intelligent Despatching sensor works, but the controls dont.

<img width="251" alt="image" src="https://github.com/BottlecapDave/HomeAssistant-OctopusEnergy/assets/27414924/65371822-0a3f-4a68-a972-7f5beb2a8354">

**Reply by BottlecapDave** [reply]
Looks like the settings are now providing seconds, probably to be iso compliant so it's failing to load all of the settings. I'll try and release a new beta at some point soon.

Regarding missing vehicle information, octopus have no plans on exposing it and state it's not much use, but I believe other intergarions rely on this data.

**Reply by Cameleon99** [reply]
Yeah that makes sense with the settings. 

Shame about the car details, but I don't use it for anything personally and can kinda see their point. It's still nice to know what make, model and most importantly battery size they have on file for you to know it's right. I could see integrations wanting to use that to do calculations. My own automation just takes a 'desired' SoC (from a helper) and sees what the car current SoC is via the Renault integration and sets the 'percentage to add' on your integration based on it. So no need to know the battery size myself.

**Reply by BottlecapDave** [reply]
New release can be found at https://github.com/BottlecapDave/HomeAssistant-OctopusEnergy/releases/tag/intelligent-api-changes-beta-03 to hopefully fix time conversion issue.

**Reply by Cameleon99** [reply]
Seems to be working now. My automation has been able to set the percent to add and change the ready time. Schedule has come through from the OE app notification and looks good 👍

**Reply by Cameleon99** [reply]
Spoke too soon. It's updating the percentage to charge and ready time ok, but the status of the Smart Charge (I think) is wrong. I have a valid schedule set in OE app but Smart Charge is showing as off?

![image](https://github.com/BottlecapDave/HomeAssistant-OctopusEnergy/assets/27414924/2c821d48-86bf-42d6-96d8-661a51e55ffd)



![image](https://github.com/BottlecapDave/HomeAssistant-OctopusEnergy/assets/27414924/ebf1b9aa-55f5-4428-91a8-2bd5c51cb478)


Can't see anything untoward in the logs. Any idea what to look for?

**Reply by BottlecapDave** [reply]
I think I'm misinterpreting the smart charge flag - it should be on all of the time unless you've explicitly disabled control. I've released https://github.com/BottlecapDave/HomeAssistant-OctopusEnergy/releases/tag/intelligent-api-changes-beta-04 which should fix this issue.

**Reply by Cameleon99** [reply]
Spot on, the Smart Charge status is showing if OE can control the device so should be on unless you've disabled the control (pretty hidden away in the app). 

Working as expected now 👍

I'm probably in the minority as I switch it on and off using my automation because I leave it plugged in most of the time, and if I don't disable it, OE tries to create a schedule each evening even when the car doesn't need it.

**Reply by BottlecapDave** [reply]
No bother. Want everything to continue working as expected :)

**Comment by BottlecapDave** [comment]
@springfall2008 - Didn't know how else to alert you to this. OE are deprecating some of the existing intelligent APIs in favour of new ones. The side effect of this is that users won't get both vehicle battery size and charger power, only one of them depending on how they're integrated (e.g. tesla communicates with the car to my knowledge, but Zappi communicates with the charger) - see https://github.com/BottlecapDave/HomeAssistant-OctopusEnergy/discussions/921#discussioncomment-9815914.

I've raised this in the [forums](https://forum.octopus.energy/t/planneddispatches-and-completeddispatches-query-changes/9735/10?u=bottlecapdave), but they currently have no plans of exposing both values

> This isn’t currently exposed but the values are not as useful as you may think beyond registration.
They are simply used as an initial starting point and best guess. The system learns and adjusts its internal view of these, as customers often pick the wrong options.

I know that these were included in this integration as part of request of your predbat integration, but not sure how you use the info. If you're needing both, it would be good to propose a case to OE.
